### DISCUSSION TITLE
other interesting projects

### DISCUSSION THREAD

**Discussion by hrgdavor** [jsxcad/JSxCAD#1176] [Show and tell]
https://discourse.threejs.org/t/three-bvh-csg-a-library-for-performing-fast-csg-operations/42713

claims even 100x performance for booleans

look at this smooth realtime boolean :)

https://gkjohnson.github.io/three-bvh-csg/examples/bundle/simple.html

It is dependant on threejs, but is interesting to explore nevertheless
https://github.com/gkjohnson/three-bvh-csg

I was not sure what would be a place if any for posting something like this..

**Comment by pentacular** [comment]
Bounding Volume Hierarchy is a nice strategy for spatial decomposition, but it doesn't address the fundamental problem of producing manifold output.

Still, these libraries look good for accelerating display where manifold output isn't necessary.

This is something that I hope to look into more in the future -- but it may require some significant architectural changes.

Thanks for sharing -- I appreciate the discovery. :)

**Comment by BarbourSmith** [comment]
What are the advantages and disadvantages of being tightly coupled with CGAL? If a better engine turned up would we ever consider switching?

**Reply by hrgdavor** [reply]
advantages of abstracting modeling part from the library is to be able to switch engines, but then also you get limited by the lowest common denominator. 
I It is something I plan to experiment with a bit, as there are now at maybe even 5 engines in my sights ... but I might get to the that point in few years after now, based on all other nifty things I would like to try :D :D :D

**Comment by pentacular** [comment]
It doesn't require abstraction, per se, but rather the ability to translate between representations, and the preservation of fundamental invariants.

For example, we already use manifold for boolean operations where exact results aren't required.
Manifold booleans have the advantage that they preserves manifold structure and do not introduce self-intersections.

We also have translations with threejs representations, so integrating this wouldn't be technically particularly challenging.
You'd just translate, do the operations, and translate back again.

I think there are two main questions that I'd want to be able to answer:
1. How much would it cost to repair the outputs so that they could be used as subsequent inputs?
2. What would be the actual performance benefits for the usual use-cases?

My guess is that (1) would be prohibitively expensive, and that (2) would mostly benefit for complex scenes where spatial decomposition would help significantly.

Given that, I'd expect this to be something that would make sense to use on the end-results -- in display.

For example, we do ray-casting to position a cursor on the surface of the object under the mouse position to show the normal position.
That could potentially benefit from this approach, although so far I haven't noticed any performance issue with the current implementation.

And we allow shapes to be moved around in the display. This approach could apply boolean operations in the display where they could be seen but not depended on by subsequent operations, and that might be sped up by this approach. Although this would be complicated by cases where earlier booleans would need to be undone, so that might be tricky.

The most reasonable use-case that comes to mind is in visualization tools on the end result -- for example, allowing dynamic cut-aways to be computed as in the demo shown.

Or in interactive construction in the display which is later fed back into the design process.

Of course, if they produce watertight manifold output that would make a big difference.

I think probably that the biggest take-away here should be that the idea of "switching engines" isn't a very useful one, but rather we can have alternatives which can be selected between for various uses, and these alternatives can have qualitative differences.

e.g., manifold is faster and manifold-proof, but vulnerable to facet cracking and shifting. cgal is slower and manifold-proof and preserves facet-level integrity.
